rita/src/rita_common/peer_listener/message.rs

```diff
@@ -0,0 +1,208 @@
+    /**
+     * Encode an ImHere message
+     * Message format is very simple
+     * Magic <u8>, Size <u16>, Ipaddr <u128>
+     */
+    pub fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+
+        match *self {
+            PeerMessage::ImHere(addr) => {
+                buf.put_u8(MSG_IM_HERE);
+                buf.put_u16_be(MSG_IM_HERE_LEN);
+                let ipaddr_bytes: [u8; 16] = addr.octets();
+                for i in 0..16 {
+                    buf.put_u8(ipaddr_bytes[i]);
+                }
+                trace!("Encoded ImHere packet {:x?}", buf);
+                return buf;
+            }
+        }
+    }
+    /**
+     * Decode buffer of data into a ImHere message
+     * Message format is very simple
+     * Magic <u8>, Size <u16>, Ipaddr <u128>
+     */
+    pub fn decode(buf: &Vec<u8>) -> Result<PeerMessage, MessageError> {
+        trace!("Starting ImHere packet decode!");
+        // Check if buffer is empty
+        if buf.is_empty() {
+            trace!("Recieved an empty ImHere packet!");
+            return Err(MessageError::InvalidPayloadError);
+        }
+        let mut pointer = Cursor::new(&buf);
+        let packet_magic = pointer.read_u8()?;
+
+        match packet_magic {
+            MSG_IM_HERE => {
+                let packet_size = pointer.read_u16::<BigEndian>()?;
+                if packet_size < MSG_IM_HERE_LEN {
+                    trace!(
+                        "Recieved an ImHere packet with an invalid size: {:?}",
+                        packet_size
+                    );
+                    return Err(MessageError::BufferUnderflow);
+                }
+
+                let mut peer_address_arr: [u16; 8] = [0xFFFF; 8];
+                for i in (0..8).rev() {
+                    peer_address_arr[i] = pointer.read_u16::<BigEndian>()?;
+                }
+                let peer_address = Ipv6Addr::new(
+                    peer_address_arr[7],
+                    peer_address_arr[6],
+                    peer_address_arr[5],
+                    peer_address_arr[4],
+                    peer_address_arr[3],
+                    peer_address_arr[2],
+                    peer_address_arr[1],
+                    peer_address_arr[0],
+                );
+
+                if peer_address.is_unspecified()
+                    || peer_address.is_loopback()
+                    || peer_address.is_multicast()
+                {
+                    trace!(
+                        "Recieved a valid ImHere with an invalid ip address: {:?}",
+                        peer_address,
+                    );
+                    return Err(MessageError::InvalidIpAddress);
+                }
+
+                trace!("ImHere decoding completed successfully {:?}", peer_address);
+                Ok(PeerMessage::ImHere(peer_address))
+            }
+            _ => {
+                trace!(
+                    "Recieved an ImHere packet with an invalid magic: {:X?}",
```

Comment:
What makes it ImHere specifically if the magic is wrong? Say that it's just an unknown magic